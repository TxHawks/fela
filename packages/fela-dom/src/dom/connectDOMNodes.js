/* @flow */
import objectEach from 'fast-loops/lib/objectEach'
import arrayEach from 'fast-loops/lib/arrayEach'

import {
  clusterCache,
  cssifySupportRules,
  sheetMap,
  RULE_TYPE,
} from 'fela-utils'

import initDOMNode from './initDOMNode'
import findDOMNodes from './findDOMNodes'

import type { DOMRenderer } from '../../../../flowtypes/DOMRenderer'

export default function connectDOMNodes(renderer: DOMRenderer): void {
  renderer.nodes = findDOMNodes()

  const cacheCluster = clusterCache(
    renderer.cache,
    renderer.mediaQueryOrder,
    renderer.supportQueryOrder
  )

  const baseNode = renderer.nodes[RULE_TYPE]

  objectEach(sheetMap, (type, key) => {
    if (cacheCluster[key].length > 0) {
      initDOMNode(renderer.nodes, baseNode, cacheCluster[key], type)
    }
  })

  const support = cssifySupportRules(cacheCluster.supportRules)

  if (support) {
    initDOMNode(renderer.nodes, baseNode, support, RULE_TYPE, '', true)
  }

  const mediaKeys = Object.keys({
    ...cacheCluster.supportMediaRules,
    ...cacheCluster.mediaRules,
  })

  arrayEach(mediaKeys, media => {
    if (
      cacheCluster.mediaRules[media] &&
      cacheCluster.mediaRules[media].length > 0
    ) {
      initDOMNode(
        renderer.nodes,
        baseNode,
        cacheCluster.mediaRules[media],
        RULE_TYPE,
        media
      )
    }

    if (cacheCluster.supportMediaRules[media]) {
      const mediaSupport = cssifySupportRules(
        cacheCluster.supportMediaRules[media]
      )

      if (mediaSupport.length > 0) {
        initDOMNode(
          renderer.nodes,
          baseNode,
          mediaSupport,
          RULE_TYPE,
          media,
          true
        )
      }
    }
  })
}
