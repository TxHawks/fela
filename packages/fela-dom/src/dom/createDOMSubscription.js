/* @flow */
/* eslint-disable consistent-return */
import objectEach from 'fast-loops/lib/objectEach'
import {
  RULE_TYPE,
  KEYFRAME_TYPE,
  FONT_TYPE,
  STATIC_TYPE,
  CLEAR_TYPE,
  generateCSSRule,
} from 'fela-utils'

import getDOMNode from './getDOMNode'
import generateRule from './generateRule'

const changeHandlers = {
  [RULE_TYPE]: (node, { selector, declaration, support }) => {
    const cssRule = generateRule(selector, declaration, support)

    // only use insertRule in production as browser devtools might have
    // weird behavior if used together with insertRule at runtime
    if (process.env.NODE_ENV !== 'production') {
      node.textContent += cssRule
      return
    }

    try {
      node.sheet.insertRule(cssRule, node.sheet.cssRules.length)
    } catch (e) {
      // TODO: warning?
    }
  },
  [KEYFRAME_TYPE]: (node, { keyframe }) => {
    node.textContent += keyframe
  },
  [FONT_TYPE]: (node, { fontFace }) => {
    node.textContent += fontFace
  },
  [STATIC_TYPE]: (node, { selector, css }) => {
    if (selector) {
      node.textContent += generateCSSRule(selector, css)
    } else {
      node.textContent += css
    }
  },
}

export default function createDOMSubscription(nodes: Object): Function {
  const baseNode = nodes[RULE_TYPE]

  return function changeSubscription(change) {
    if (change.type === CLEAR_TYPE) {
      return objectEach(nodes, node => {
        node.textContent = ''
      })
    }

    const handleChange = changeHandlers[change.type]

    if (handleChange) {
      const node = getDOMNode(
        nodes,
        baseNode,
        change.type,
        change.media,
        !!change.support
      )

      handleChange(node, change)
    }
  }
}
