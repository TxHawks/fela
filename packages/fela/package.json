{
  "name": "fela",
  "version": "6.0.0",
  "description": "Universal, Dynamic & High-Performance Styling in JavaScript",
  "typings": "index.d.ts",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "scripts": {
    "postinstall": "opencollective postinstall"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.d.ts",
    "lib/**",
    "es/**"
  ],
  "repository": "https://github.com/rofrischmann/fela/",
  "keywords": [
    "fela",
    "dynamic styling",
    "stylesheet",
    "css",
    "styling",
    "cssinjs"
  ],
  "author": "Robin Frischmann",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "^2.0.0",
    "fela-utils": "^7.0.0",
    "opencollective": "^1.0.3"
  },
  "devDependencies": {
    "fela-tools": "^5.0.8"
  },
  "collective": {
    "type": "opencollective",
    "url": "https://opencollective.com/fela",
    "logo": "https://opencollective.com/opencollective/logo.txt"
  }
}
