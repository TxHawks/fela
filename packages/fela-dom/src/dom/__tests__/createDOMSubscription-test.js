import { html as beautify } from 'js-beautify'
import { createRenderer } from 'fela'

import createDOMSubscription from '../createDOMSubscription'
import connectDOMNodes from '../connectDOMNodes'

beforeEach(() => {
  const head = document.head
  while (head.firstChild) {
    head.removeChild(head.firstChild)
  }
})

afterEach(() => {
  process.env.NODE_ENV = 'test'
})

describe('Subscribing to the DOM', () => {
  it('should render rules to a DOM node', () => {
    const renderer = createRenderer()

    connectDOMNodes(renderer)

    const updateSubscription = createDOMSubscription(renderer.nodes)
    renderer.subscribe(updateSubscription)

    renderer.renderRule(() => ({
      backgroundColor: 'red',
      color: 'blue',
      ':hover': {
        color: 'red',
      },
    }))

    expect(
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      })
    ).toMatchSnapshot()
  })

  it('should render static styles to a DOM node', () => {
    const renderer = createRenderer()

    connectDOMNodes(renderer)

    const updateSubscription = createDOMSubscription(renderer.nodes)
    renderer.subscribe(updateSubscription)

    renderer.renderStatic(
      {
        backgroundColor: 'red',
        color: 'blue',
      },
      'body, html'
    )

    expect(
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      })
    ).toMatchSnapshot()
  })

  it('should render keyframes to a DOM node', () => {
    const renderer = createRenderer()

    connectDOMNodes(renderer)

    const updateSubscription = createDOMSubscription(renderer.nodes)
    renderer.subscribe(updateSubscription)

    renderer.renderKeyframe(() => ({
      from: { color: 'red' },
      to: { color: 'blue' },
    }))

    expect(
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      })
    ).toMatchSnapshot()
  })

  it('should render media rules and support rules to single DOM nodes', () => {
    const renderer = createRenderer()

    connectDOMNodes(renderer)

    const updateSubscription = createDOMSubscription(renderer.nodes)
    renderer.subscribe(updateSubscription)

    renderer.renderRule(() => ({
      color: 'blue',
      '@supports (display: flex)': {
        display: 'flex',
      },
      '@media (min-width: 300px)': {
        color: 'red',
        '@media (max-height: 500px)': {
          color: 'yellow',
          '@supports (display: grid)': {
            display: 'grid',
          },
        },
      },
    }))

    expect(
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      })
    ).toMatchSnapshot()
  })

  it('should clear all DOM nodes', () => {
    const renderer = createRenderer()

    connectDOMNodes(renderer)

    const updateSubscription = createDOMSubscription(renderer.nodes)
    renderer.subscribe(updateSubscription)

    renderer.renderRule(() => ({
      color: 'blue',
      '@media (min-width: 300px)': {
        color: 'red',
        '@media (max-height: 500px)': {
          color: 'yellow',
        },
      },
    }))

    renderer.renderKeyframe(() => ({
      from: { color: 'red' },
      to: { color: 'blue' },
    }))

    renderer.clear()

    expect(
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      })
    ).toMatchSnapshot()
  })

  it('should use insertRule', () => {
    const renderer = createRenderer()

    process.env.NODE_ENV = 'production'

    connectDOMNodes(renderer)

    const updateSubscription = createDOMSubscription(renderer.nodes)
    renderer.subscribe(updateSubscription)

    renderer.renderRule(() => ({
      color: 'blue',
      '@media (min-width: 300px)': {
        color: 'red',
        '@media (max-height: 500px)': {
          color: 'yellow',
        },
      },
    }))

    renderer.renderKeyframe(() => ({
      from: { color: 'red' },
      to: { color: 'blue' },
    }))

    const rules = {}

    document.querySelectorAll('[data-fela-type="RULE"]').forEach(node => {
      const sheet = node.sheet
      const media = node.getAttribute('media') || ''

      rules[media] = sheet.cssRules
    })

    expect([
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      }),
      rules,
    ]).toMatchSnapshot()
  })
})
