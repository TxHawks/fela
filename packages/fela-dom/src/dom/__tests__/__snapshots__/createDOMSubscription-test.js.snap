// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribing to the DOM should clear all DOM nodes 1`] = `
"<html>

<head>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\"></style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" media=\\"(min-width: 300px)\\"></style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" media=\\"(min-width: 300px) and (max-height: 500px)\\"></style>
  <style data-fela-type=\\"KEYFRAME\\" type=\\"text/css\\"></style>
</head>

<body></body>

</html>"
`;

exports[`Subscribing to the DOM should render keyframes to a DOM node 1`] = `
"<html>

<head>
  <style data-fela-type=\\"KEYFRAME\\" type=\\"text/css\\">
    @-webkit-keyframes k1 {
      from {
        color: red
      }
      to {
        color: blue
      }
    }

    @-moz-keyframes k1 {
      from {
        color: red
      }
      to {
        color: blue
      }
    }

    @keyframes k1 {
      from {
        color: red
      }
      to {
        color: blue
      }
    }
  </style>
</head>

<body></body>

</html>"
`;

exports[`Subscribing to the DOM should render media rules and support rules to single DOM nodes 1`] = `
"<html>

<head>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\">
    .a {
      color: blue
    }
  </style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" data-fela-support=\\"true\\">
    @supports (display: flex) {
      .b {
        display: flex
      }
    }
  </style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" media=\\"(min-width: 300px)\\">
    .c {
      color: red
    }
  </style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" media=\\"(min-width: 300px) and (max-height: 500px)\\">
    .d {
      color: yellow
    }
  </style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" data-fela-support=\\"true\\" media=\\"(min-width: 300px) and (max-height: 500px)\\">
    @supports (display: grid) {
      .e {
        display: grid
      }
    }
  </style>
</head>

<body></body>

</html>"
`;

exports[`Subscribing to the DOM should render rules to a DOM node 1`] = `
"<html>

<head>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\">
    .a {
      background-color: red
    }

    .b {
      color: blue
    }

    .c:hover {
      color: red
    }
  </style>
</head>

<body></body>

</html>"
`;

exports[`Subscribing to the DOM should render static styles to a DOM node 1`] = `
"<html>

<head>
  <style data-fela-type=\\"STATIC\\" type=\\"text/css\\">
    body,
    html {
      background-color: red;
      color: blue
    }
  </style>
</head>

<body></body>

</html>"
`;

exports[`Subscribing to the DOM should use insertRule 1`] = `
Array [
  "<html>

<head>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\"></style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" media=\\"(min-width: 300px)\\"></style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" media=\\"(min-width: 300px) and (max-height: 500px)\\"></style>
  <style data-fela-type=\\"KEYFRAME\\" type=\\"text/css\\">
    @-webkit-keyframes k1 {
      from {
        color: red
      }
      to {
        color: blue
      }
    }

    @-moz-keyframes k1 {
      from {
        color: red
      }
      to {
        color: blue
      }
    }

    @keyframes k1 {
      from {
        color: red
      }
      to {
        color: blue
      }
    }
  </style>
</head>

<body></body>

</html>",
  Object {
    "": Array [
      CSSStyleRule {
        "__ends": 14,
        "__starts": 0,
        "parentRule": null,
        "parentStyleSheet": CSSStyleSheet {
          "cssRules": [Circular],
          "parentStyleSheet": null,
        },
        "selectorText": ".a",
        "style": CSSStyleDeclaration {
          "0": "color",
          "__starts": 2,
          "_importants": Object {
            "color": "",
          },
          "color": "blue",
          "length": 1,
          "parentRule": [Circular],
        },
      },
    ],
    "(min-width: 300px)": Array [
      CSSStyleRule {
        "__ends": 13,
        "__starts": 0,
        "parentRule": null,
        "parentStyleSheet": CSSStyleSheet {
          "cssRules": [Circular],
          "parentStyleSheet": null,
        },
        "selectorText": ".b",
        "style": CSSStyleDeclaration {
          "0": "color",
          "__starts": 2,
          "_importants": Object {
            "color": "",
          },
          "color": "red",
          "length": 1,
          "parentRule": [Circular],
        },
      },
    ],
    "(min-width: 300px) and (max-height: 500px)": Array [
      CSSStyleRule {
        "__ends": 16,
        "__starts": 0,
        "parentRule": null,
        "parentStyleSheet": CSSStyleSheet {
          "cssRules": [Circular],
          "parentStyleSheet": null,
        },
        "selectorText": ".c",
        "style": CSSStyleDeclaration {
          "0": "color",
          "__starts": 2,
          "_importants": Object {
            "color": "",
          },
          "color": "yellow",
          "length": 1,
          "parentRule": [Circular],
        },
      },
    ],
  },
]
`;
